File: 03.CPP/0053.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int maxSubArray(std::vector<int>& nums) {
        int bestLocalSum = nums[0];
        int bestGlobalSum = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            bestLocalSum = std::max(nums[i], nums[i] + bestLocalSum);
            bestGlobalSum = std::max(bestGlobalSum, bestLocalSum);
        }

        return bestGlobalSum;
    }
};

int main() {
    Solution sol;

    std::vector<int> nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    int result = sol.maxSubArray(nums);

    std::cout << result << "\n";
    return 0;
}

File: 01.Python/0053.py
from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        localSum = nums[0]
        globalBestSum = nums[0]

        for num in nums[1:]:
            localSum = max(num, num + localSum)
            globalBestSum = max(globalBestSum, localSum)
        
        return globalBestSum

sol = Solution()

nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
print(sol.maxSubArray(nums))